File: pset6/credit/credit.py
from cs50 import get_string

# Get card number that is at most 16 characters long
while True:
    card = get_string("Number: ")
    if len(card) <= 16:
        break

# Initialize variables
length = len(card) - 1
total = 0

while True:
    if length < 1:
        break

    # Take every other number from the back and multiply by 2
    number = str(int(card[length - 1])* 2)

    # Manipulate strings into int
    if len(number) > 1:
        total += int(number[0]) + int(number[1]) + int(card[length])
    else:
        total += int(number[0]) + int(card[length])

    # Do in pairs
    length -= 2

# Add the first number of len() is odd
if len(card) % 2 == 1:
    total += int(card[0])

# Check for algorithm
if total % 10 != 0:
    print("INVALID")
elif len(card) == 15 and card[0] == "3" and (card[1] == "4" or "7"):
    print("AMEX")
elif len(card) == 16 and card[0] == "5" and int(card[1]) < 6:
    print("MASTERCARD")
elif (len(card) == 13 or 16) and card[0] == "4":
    print("VISA")
else:
    print("INVALID")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

long getcard(void); //Prompt for card within range
int getdigit(long card, int X); //Get (X)th digit from the right

int main(void)
{
    long card = getcard(); //Prompt for card within range


    int j, doubleeven, even = 0; //Gather every other digit
    for (j = 1; j < 17; j++)
    {
        if (j % 2 == 0) //Make sure they are 'every other'
        {
            doubleeven = 2 * getdigit(card, j);
            even += (getdigit(doubleeven, 1) + getdigit(doubleeven, 2)); //Split them and sum
        }
    }

    int i, odd = 0; //Gather the odd digits
    for (i = 1; i < 17; i++)
    {
        if (i % 2 == 1) //Make sure they are odd
        {
            odd += getdigit(card, i); //Sum the odds
        }
    }

    int checksum = even + odd;
    if (checksum % 10 != 0) //Check for Luhn's Algorithm
    {
        printf("INVALID\n");
    }
    else if (getdigit(card, 15) == 3 && ((getdigit(card, 14) == 4) || (getdigit(card, 14) == 7))) //Check for AMEX
    {
        printf("AMEX\n");
    }
    else if (getdigit(card, 16) == 5 && (getdigit(card, 15) < 6)) //Check for MASTERCARD
    {
        printf("MASTERCARD\n");
    }
    else if (getdigit(card, 16) == 4 || getdigit(card, 13) == 4) //Check for VISA
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

long getcard(void)
{
    long card; // Prompt for a card number within 16 digits
    do
    {
        card = get_long("Card number: ");
    }
    while (card > 9999999999999999); // Card does not exceed 16 characters
    return card;
}

int getdigit(long card, int X) //Function to show (X)th number from the right
{
    long tenpow = 1;
    for (int i = 1; i < X; i++) //X is the (X)th number from the right
    {
        tenpow *= 10;
    }
    card /= tenpow; //Reduce card number to (X)
    card %= 10; //Get last digit
    return card;
}